File: docs/test-cases/oauth-implementation.java
// Test case: oauth-implementation (A07:2025)
package com.example.auth;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jws;
import io.jsonwebtoken.Jwts;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.view.RedirectView;

@RestController
public class OAuthController {

    private static final String TRUSTED_BASE = "https://app.com";
    private static final String SIGNING_KEY = "hardcoded_signing_key_xyz";

    @GetMapping("/oauth/start")
    public RedirectView start(@RequestParam String redirect_uri) {
        // BUG: no `state` parameter generated — OAuth callback has no CSRF protection
        String url = "https://idp.example.com/auth?client_id=myapp&redirect_uri=" + redirect_uri;
        return new RedirectView(url);
    }

    @GetMapping("/oauth/callback")
    public String callback(@RequestParam String code,
                           @RequestParam String redirect_uri,
                           @RequestParam String token) {
        // BUG: prefix match — "https://app.com.attacker.io/steal" passes
        if (!redirect_uri.startsWith(TRUSTED_BASE)) {
            return "bad redirect";
        }

        // BUG: no `state` parameter validated against session — CSRF
        // BUG: Jwts.parser().setSigningKey(...) accepts alg:none unsigned tokens
        //      and never validates the `aud` claim
        Jws<Claims> parsed = Jwts.parser()
                .setSigningKey(SIGNING_KEY.getBytes())
                .parseClaimsJws(token);
        Claims claims = parsed.getBody();

        String userId = claims.getSubject();
        return "welcome " + userId;
    }
}

File: docs/test-cases/oauth-implementation.go
// Test case: oauth-implementation (A07:2025)
package auth

import (
	"net/http"
	"strings"

	"github.com/golang-jwt/jwt/v5"
)

const trustedBase = "https://app.example.com"

var signingKey = []byte("hardcoded_hs256_secret")

func OAuthStart(w http.ResponseWriter, r *http.Request) {
	redirectURI := r.URL.Query().Get("redirect_uri")
	// BUG: no `state` generated — callback cannot detect CSRF
	http.Redirect(w, r,
		"https://idp.example.com/auth?client_id=myapp&redirect_uri="+redirectURI,
		http.StatusFound)
}

func OAuthCallback(w http.ResponseWriter, r *http.Request) {
	redirectURI := r.URL.Query().Get("redirect_uri")
	// BUG: HasPrefix lets "https://app.example.com.attacker.io/x" through
	if !strings.HasPrefix(redirectURI, trustedBase) {
		http.Error(w, "bad redirect", 400)
		return
	}

	tokenStr := r.URL.Query().Get("token")
	// BUG: Keyfunc never checks token.Method — alg:none and HS/RS confusion both pass
	tok, err := jwt.Parse(tokenStr, func(t *jwt.Token) (interface{}, error) {
		return signingKey, nil
	})
	if err != nil || !tok.Valid {
		http.Error(w, "bad token", 401)
		return
	}
	// BUG: no `aud` claim check, no `state` validated against session

	// BUG: token written to a cookie without Secure or HttpOnly — XSS-readable
	http.SetCookie(w, &http.Cookie{
		Name:  "access_token",
		Value: tokenStr,
		Path:  "/",
	})
	w.Write([]byte("ok"))
}
